File: bonus/c/data_structures/stack.c
/*
  Stack data structure example
  Copyright 2017, Sjors van Gelderen
*/

#include <stdlib.h>
#include <stdio.h>

// Stack data structure
typedef struct Stack {
    int size;
    int capacity;
    int* values;
} Stack;

// Function prototypes
void push(Stack* _stack, int _value);
void pop(Stack* _stack);
int peek(Stack* _stack);

/*
  Push an element onto the stack
  Complexity: O(1)
*/
void push(Stack* _stack, int _value)
{
    if(_stack->capacity > 0)
    {
	printf("PUSH: %d\n", _value);
	_stack->values[_stack->capacity - 1] = _value;
	_stack->capacity--;
    }
    else
    {
	// Stack capacity reached
	printf("PUSH: %d - Error: stack overflow!\n", _value);
    }
}

/*
  Pop an element from the stack
  Complexity: O(1)
*/
void pop(Stack* _stack)
{
    if(_stack->capacity < _stack->size)
    {
	printf("POP\n");
	_stack->capacity++;
    }
    else
    {
	// No elements on stack to pop
	printf("Error: stack underflow!\n");
    }
}

/*
  Peek the top element of the stack
  Complexity: O(1)
*/
int peek(Stack* _stack)
{
    if(_stack->capacity == _stack->size)
    {
	// No elements on stack to peek
	printf("Error: invalid operation!\n");
	return -1;
    }
    else
    {
	int value = _stack->values[(_stack->size - 1) - _stack->capacity];
	printf("PEEK: %d\n", value);
	return value;
    }
}

int main()
{
    printf("Stack data structure example - ");
    printf("Copyright 2017, Sjors van Gelderen\n\n");
    
    const int SIZE = 4;

    // Create a stack
    Stack stack;
    stack.size = SIZE;
    stack.capacity = SIZE;
    stack.values = malloc(sizeof(int) * SIZE);

    // Do some operations on it
    push(&stack, 1);
    push(&stack, 2);
    push(&stack, 3);
    push(&stack, 4);
    push(&stack, 5);

    peek(&stack);

    pop(&stack);
    pop(&stack);
    
    peek(&stack);

    pop(&stack);
    pop(&stack);
    pop(&stack);

    peek(&stack);

    // Free and deallocate memory
    free(stack.values);
    stack.values = NULL;
    
    return EXIT_SUCCESS;
}

File: bonus/fsharp/data_structures/stack.fs
(*
Stack data structure example
Copyright 2016, Sjors van Gelderen
*)

module Program

//Stack data structure
type Stack<'a> =
    { Size     : int
      Segments : 'a List } with
    static member Zero =
        { Size = 0
          Segments = List.empty }

(*
Push an element onto the stack
Complexity: O(1)
*)
let push element stack =
    if stack.Segments.Length < stack.Size then
        printfn "PUSH: %A" element
        { stack with Segments = element :: stack.Segments }
    else
        printfn "PUSH: Stack overflow"
        stack

(*
Pop an element from the stack
Complexity: O(1)
*)
let pop stack =
    if stack.Segments.Length > 0 then
        printfn "POP: %A" stack.Segments.Head
        { stack with Segments = stack.Segments.Tail }
    else
        printfn "POP: Stack underflow"
        stack

(*
Peek the top element of the stack
Complexity: O(1)
*)
let peek stack =
    if stack.Segments.IsEmpty then
        printfn "PEEK: Stack underflow"
    else
        printfn "PEEK: %A" stack.Segments.Head

(*
Print all elements in the stack
Complexity: O(n)
*)
let print stack =
    printfn "REVEAL:"
    List.iter (fun elem -> printfn "%A" elem) stack.Segments

// Main program logic
[<EntryPoint>]
let main args =
    printfn "Stack data structure example - \
             Copyright 2016, Sjors van Gelderen"

    printfn "Stack with size 4"
    let populate = List.fold (fun acc elem -> push elem acc)
    let stack = populate { Stack<int>.Zero with Size = 4 } [ 1; 2; 3; 4; 5 ]
    print stack
    
    let stack = pop stack
    peek stack
    
    let stack = stack |> pop |> pop
    peek stack
    print stack

    let stack = stack |> push 10 |> push 20
    print stack
    
    0
